### PR TITLE
Compact rcf

### PR DESCRIPTION
**Pull request by sudiptoguha**
*Issue #, if available:* 148 (and exposing conditional fields)

*Description of changes:* removes the bug in bounding box computation after deserialization -- all of the bounding boxes are not computed for efficiency reasons.

### CODE REVIEW

File: Java/core/src/main/java/com/amazon/randomcutforest/RandomCutForest.java [hunk 1]
```diff
@@ -815,24 +817,20 @@ public DensityOutput getSimpleDensity(double[] point) {
      *                              missing values.
      * @return A point with the missing values imputed.
      */
-    public double[] imputeMissingValues(double[] point, int numberOfMissingValues, int[] missingIndexes) {
-        checkArgument(numberOfMissingValues >= 0, "numberOfMissingValues must be greater than or equal to 0");
+    public ArrayList<?> getSimpleConditionalField(double[] point, int numberOfMissingValues, int[] missingIndexes) {
```
Comment (jotok) [inline Java/core/src/main/java/com/amazon/randomcutforest/RandomCutForest.java:820]: I think it's a little ugly to have `ArrayList<?>` as the return type in this public API, requiring the user to cast the result in order to use it. We could have this method return `ArrayList<double[]>`, where the arrays would be single-element arrays when number of missing values is 1.

File: Java/core/src/test/java/com/amazon/randomcutforest/CompactRandomCutForestFunctionalTest.java [hunk 2]
```diff
@@ -866,4 +868,30 @@ public void getTotalUpdates_returnExpectedSize() {
         }
         return shingledPoint;
     }
+
+    @Test
+    public void reproduce() {
```
Comment (jotok) [inline Java/core/src/test/java/com/amazon/randomcutforest/CompactRandomCutForestFunctionalTest.java:873]: Let's call this something like `testUpdateAfterRoundTrip()`.

Comment (sudiptoguha) [reply Java/core/src/test/java/com/amazon/randomcutforest/CompactRandomCutForestFunctionalTest.java:873]: Done

File: Java/core/src/main/java/com/amazon/randomcutforest/RandomCutForest.java (hunk 1 above)
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/RandomCutForest.java:820]: done

File: Java/core/src/main/java/com/amazon/randomcutforest/RandomCutForest.java [hunk 3]
```diff
@@ -815,6 +817,36 @@ public DensityOutput getSimpleDensity(double[] point) {
      *                              missing values.
      * @return A point with the missing values imputed.
      */
+    public ArrayList<double[]> getSimpleConditionalField(double[] point, int numberOfMissingValues,
```
Comment (jotok) [inline Java/core/src/main/java/com/amazon/randomcutforest/RandomCutForest.java:820]: I think we need a better name for this method, but we can do it in another PR.

#151

**Review by jotok — APPROVED** [review]
